### PR TITLE
feat: [CDS-72501]: gitops revertPR step

### PR DESCRIPTION
**Pull request by manavjot-harness**
## Describe your changes
Changes for adding a new revert PR step for the gitops pipeline flow.
This step reverts the commit done by the update release repo step if required based on the user's usecase.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50130)

### CODE REVIEW

**Comment by manavjot-harness** [comment]
trigger feature-build

**Comment by manavjot-harness** [comment]
trigger feature-build

**Comment by manavjot-harness** [comment]
trigger feature-build

**Comment by manavjot-harness** [comment]
trigger feature-build

**Comment by manavjot-harness** [comment]
trigger feature-build

**Comment by manavjot-harness** [comment]
trigger feature-build

**Comment by manavjot-harness** [comment]
trigger publish-delegate

**Comment by manavjot-harness** [comment]
trigger publish-delegate

**Comment by manavjot-harness** [comment]
trigger feature-build

**Comment by manavjot-harness** [comment]
trigger feature-build

**Comment by manavjot-harness** [comment]
trigger publish-delegate

File: 125-cd-nextgen/src/main/java/io/harness/cdng/gitops/revertpr/RevertPRStep.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+      throw new InvalidRequestException(String.format("Failed to execute Revert PR step. %s", e.getMessage()));
+    }
+  }
+
+  @Override
+  public Class<StepElementParameters> getStepParametersClass() {
+    return null;
+  }
+
+  public GitFetchFilesConfig getGitFetchFilesConfig(
```
Comment (lasdox) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/gitops/revertpr/RevertPRStep.java:162]: I wonder if we should change this method name, doesn't make much sense for revertPR context

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java [hunk 2]
```diff
@@ -342,6 +383,49 @@ public DelegateResponseData handleUpdateReleaseRepo(NGGitOpsTaskParams gitOpsTas
...
   }
 
+  public DelegateResponseData handleRevertPR(NGGitOpsTaskParams gitOpsTaskParams) {
+    CommandUnitsProgress commandUnitsProgress = CommandUnitsProgress.builder().build();
+    try {
+      log.info("Running Revert PR Task for activityId {}", gitOpsTaskParams.getActivityId());
+      logCallback = new NGDelegateLogCallback(getLogStreamingTaskClient(), RevertPR, true, commandUnitsProgress);
+
+      // need this to set the auth mechanisms for the repo
+      getFilesRevert(gitOpsTaskParams);
```
Comment (lasdox) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java:393]: could you explain more on why this is needed? revert shoudnt need to parse any files.
Is it manipulating the gitOpsTaskParams? If yes it is not easy to understand what is going on.

**Review by lasdox — DISMISSED** [review]
Even though I think there are room for improvement, I also believe it is good to go

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java [hunk 3]
```diff
@@ -342,6 +383,49 @@ public DelegateResponseData handleUpdateReleaseRepo(NGGitOpsTaskParams gitOpsTas
...
+      String baseBranch = gitOpsTaskParams.getGitFetchFilesConfig().getGitStoreDelegateConfig().getBranch();
+      String newBranch = baseBranch + "_revert_" + RandomStringUtils.randomAlphabetic(12);
+      ScmConnector scmConnector =
+          gitOpsTaskParams.getGitFetchFilesConfig().getGitStoreDelegateConfig().getGitConfigDTO();
+
+      createNewBranch(scmConnector, newBranch, baseBranch);
+      RevertAndPushResult revertAndPushResult = revert(gitOpsTaskParams, REVERT_COMMIT_MSG, newBranch);
+
+      CreatePRResponse createPRResponse =
+          createPullRequest(scmConnector, newBranch, baseBranch, REVERT_PR_TITLE, gitOpsTaskParams.getAccountId());
```
Comment (mankrit-singh) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java:404]: should we allow for a custom PR title for this step as well? for Update Release Repo we use resolvePRTitle function and take necessary optional input

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java [hunk 4]
```diff
@@ -261,6 +269,39 @@ public FetchFilesResult getFiles(NGGitOpsTaskParams gitOpsTaskParams, CommandUni
     }
   }
 
+  public FetchFilesResult getFilesRevert(NGGitOpsTaskParams gitOpsTaskParams) throws IOException {
```
Comment (meenaravichandran1) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java:272]: Is it possible to keep this lite? Why do we need FetchFilesResult? Can we write a new method to just do the auth for the repo?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java (hunk 2 above)
Comment (manavjot-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java]: changed the method.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java [hunk 5]
```diff
@@ -261,6 +269,39 @@ public FetchFilesResult getFiles(NGGitOpsTaskParams gitOpsTaskParams, CommandUni
...
+    try {
+      FetchFilesResult fetchFilesResult =
+          gitOpsTaskHelper.getFetchFilesResult(gitOpsTaskParams.getGitFetchFilesConfig(),
+              gitOpsTaskParams.getAccountId(), logCallback, gitOpsTaskParams.isCloseLogStream());
+      updateFilesNotFoundWithEmptyContent(
```
[the remark this answers is not in the document]
Comment (manavjot-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java]: removed.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java (hunk 4 above)
Comment (manavjot-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java]: done.

**Comment by manavjot-harness** [comment]
trigger publish-delegate

**Comment by manavjot-harness** [comment]
trigger uts

**Comment by manavjot-harness** [comment]
trigger uts

**Comment by manavjot-harness** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java (hunk 3 above)
Comment (manavjot-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/git/ng/NGGitOpsCommandTask.java]: done.

**Comment by manavjot-harness** [comment]
trigger uts

**Review by lasdox — DISMISSED** [review]
Lgtm

**Review by meenaravichandran1 — DISMISSED** [review]

**Comment by manavjot-harness** [comment]
trigger uts

**Comment by manavjot-harness** [comment]
trigger uts

**Comment by mankrit-singh** [comment]
trigger ut7

**Comment by mankrit-singh** [comment]
trigger uts

**Comment by mankrit-singh** [comment]
trigger ut7

**Review by mankrit-singh — APPROVED** [review]

**Review by mankrit-singh — DISMISSED** [review]

**Comment by mankrit-singh** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]

**Comment by mankrit-singh** [comment]
trigger review

**Comment by mankrit-singh** [comment]
trigger ut3

**Comment by manavjot-harness** [comment]
trigger uts

**Review by mankrit-singh — APPROVED** [review]

**Review by yogesh-chauhan — APPROVED** [review]

**Comment by mankrit-singh** [comment]
trigger review

**Comment by mankrit-singh** [comment]
trigger ut8

**Comment by mankrit-singh** [comment]
trigger ut7

**Comment by manavjot-harness** [comment]
trigger ut0,ut4,ut8

**Pull request merged**